SearXNG JavaScript Style Guide

return42 opened the pull request:
## What does this PR do?

Define a SearXNG JavaScript Style Guide

The SearXNG JavaScript Style Guide is a set of eslint rules [1] based on the
eslint:recommended [2] rules.

    "extends": "eslint:recommended"

The additional rules adopted from the JavaScript Standard Style [3][4].

[1] https://eslint.org/docs/rules/
[2] https://github.com/kunalgolani/eslint-config/tree/master/packages/recommended
[3] https://standardjs.com/rules.html#javascript-standard-style
[4] https://github.com/standard/eslint-config-standard/blob/master/.eslintrc.json

The first three commits do fix small issues in preparation

## Why is this change important?

consistent code quality

return42 committed:
[emacs] .dir-locals.el: in json-mode use indentation of 4

In JS + JSON we use indentation of 4.

Signed-off-by: Markus Heiser <markus.heiser@darmarit.de>

return42 committed:
[mod] SearXNG JavaScript Style Guide (eslint:recommended)

The SearXNG JavaScript Style Guide is a set of eslint rules [1] based on the
eslint:recommended [2] rules.

    "extends": "eslint:recommended"

The additional rules adopted from the JavaScript Standard Style [3][4].

[1] https://eslint.org/docs/rules/
[2] https://github.com/kunalgolani/eslint-config/tree/master/packages/recommended
[3] https://standardjs.com/rules.html#javascript-standard-style
[4] https://github.com/standard/eslint-config-standard/blob/master/.eslintrc.json

Signed-off-by: Markus Heiser <markus.heiser@darmarit.de>

dalf commented on searx/static/themes/simple/.eslintrc.json:
```diff
@@ -9,5 +9,7 @@
         "ecmaVersion": 12
     },
     "rules": {
+        "indent": ["error", 2],
```
most of the time the value is 4 (eslint source code uses 4 space for example).
For readability, I prefer 4 (personal taste)

ping @mrpaulblack

return42 replied:
fixed / additional I added gruntfile.js itself to the files linted by eslint .. a `[build] /static` commit is also added.

mrpaulblack replied:
I really like using 2 spaces indent... But that is just because I work with a lot of yaml files when using docker-compose and writing config for traefik for example... So yeah I would prefer 2 spaces.... (But since I am not really working with the js files I leave it up to you @dalf )

return42 replied:
The comment from @mrpaulblack lead me to have a look what [JavaScript Style Guide](https://standardjs.com/rules.html) says about indent: _"Use 2 spaces for indentation."_

@dalf What do you think, should we follow the Sytle Guide?

dalf replied:
please don't merge

return42 replied:
I converted this PR to draft .. I will give [eslint-config-standard](https://github.com/standard/eslint-config-standard) a try .. coming back soon ..

return42 replied:
```
npm ERR! Found: eslint@8.3.0
npm ERR! node_modules/eslint
npm ERR!   dev eslint@"^8.0.1" from the root project
npm ERR! 
npm ERR! Could not resolve dependency:
npm ERR! peer eslint@"^7.12.1" from eslint-config-standard@16.0.3
npm ERR! node_modules/eslint-config-standard
npm ERR!   dev eslint-config-standard@"*" from the root project
npm ERR! 
```

Aargh, I am annoyed by this _nodejs version maintaining and stability_ .. we updated to ESLint v8, [eslint-config-standard](https://github.com/standard/eslint-config-standard) depends on v7 .. but there is a [PR upgrading to v8](https://github.com/standard/eslint-config-standard/pull/193) .. but this PR is blocked by its dependencies to

- https://github.com/mysticatea/eslint-plugin-node/pull/295
- https://github.com/xjamundx/eslint-plugin-promise/pull/219

And if you dive deeper in this pool off sub-sub-dependencies you will find more ESLint v7 dependencies ..

I bet my last 5cent that the day all dependencies are resolved, ESlint version 9 is released .. I feel like I'm in the movie [Groundhog Day](https://en.wikipedia.org/wiki/Groundhog_Day_(film)) 

Whatever I touch in this nodejs field is unmaintained or has version conflicts .. where is my fail?

return42 replied:
@dalf I renamed this PR to "SearXNG JavaScript Style Guide" and rewrite the initial https://github.com/searxng/searxng/pull/538#issue-787190682 for details read along the rebased commits series.

return42 committed:
[mod] eslint: add gruntfile.js to the files to lint

Signed-off-by: Markus Heiser <markus.heiser@darmarit.de>

return42 committed:
[fix] eslint --fix

automatically fix some of the problems reported by eslint rules::

    $ ./manage nvm.bash
    nvm-env$ npm --prefix searx/static/themes/simple run eslint-fix

Signed-off-by: Markus Heiser <markus.heiser@darmarit.de>

return42 committed:
[simple theme] activate - SearXNG JavaScript Style Guide

Add the 'SearXNG JavaScript Style Guide' as one quality gate in the build chain
of the simple theme::

    make themes.simple
    make themes.simple.test

In the github CI the `themes.all` target enforce a `themes.simple`.

BTW: Remove 'jshint' left overs from 0ee316f3d

Signed-off-by: Markus Heiser <markus.heiser@darmarit.de>
